### PR TITLE
Fix #6835 Big Number: Comparison of large exponentials

### PR DESCRIPTION
**Pull request by seadavis**
#### References to other Issues or PRs
Fixes #6835

#### Brief description of what is fixed or changed
Implemented a class that allows us to implement very
large exponents. There is still more work to do, but
this will at least allow us to compare number such
as 10**10**100. We cannot, in general,compare unevaluated
expressions yet, so multiplication addition and subtraction,
are just evaluated unless one contains some Power
But Pow's will be unevaluated and symbolically

#### Other comments
Known Issues:
Division between big numbers is not implemented.
there is still a maximum on the size of exponentials that
can be compared

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
I think it would be better to solve this sort of thing using #18778 and by preventing the automatic evaluation of large exponentials rather than introducing a new `BigNumber` class.

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -3413,6 +3416,216 @@ def __hash__(self):
         return super(NumberSymbol, self).__hash__()
 
 
+class BigNumber:
+    """
+
```
Comment (czgdp1807) [inline sympy/core/numbers.py:3421]:
```suggestion
```

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -3925,8 +4138,14 @@ def sympify_complex(a):
...
     return real + S.ImaginaryUnit*imag
 
+
+
 converter[complex] = sympify_complex
 
+
+
+
+
```
Comment (czgdp1807) [inline sympy/core/numbers.py:4148]:
```suggestion
```

File: sympy/core/numbers.py [hunk 3]
```diff
@@ -3925,8 +4138,14 @@ def sympify_complex(a):
     real, imag = list(map(sympify, (a.real, a.imag)))
     return real + S.ImaginaryUnit*imag
 
+
+
```
Comment (czgdp1807) [inline sympy/core/numbers.py:4142]:
```suggestion
```

File: sympy/core/tests/test_numbers.py [hunk 4]
```diff
@@ -1,4 +1,6 @@
 from __future__ import absolute_import
+
```
Comment (czgdp1807) [inline sympy/core/tests/test_numbers.py:2]:
```suggestion
```

File: sympy/core/tests/test_numbers.py [hunk 5]
```diff
@@ -1943,6 +1945,9 @@ def test_NumberSymbol_comparison():
     fpi = Float(float(pi))
     assert rel_check(rpi, fpi)
 
+
+
+
```
Comment (czgdp1807) [inline sympy/core/tests/test_numbers.py:1950]:
```suggestion
```

File: sympy/core/tests/test_numbers.py [hunk 6]
```diff
@@ -2025,3 +2030,28 @@ def test_abc():
...
+    assert (e1 < 100) == False
+    assert (100 < e2) == True
+    assert (Float(100) < e2) == True
+    assert (e2 < 100) == False
+    assert (100 < e3) == True
+    assert (e3 < e1) == True
+    assert (Rational(170, 10) < e1) == True
+    assert (Rational(170, 10) > e1) == False
+    assert (e1_prime == e1) == True
+    assert (e2 < e3) == True
```
Comment (czgdp1807) [inline sympy/core/tests/test_numbers.py:2049]:
```suggestion
    assert (100 < e1)
    assert not (e1 < 100)
    assert (100 < e2)
    assert (Float(100) < e2)
    assert not (e2 < 100)
    assert (100 < e3)
    assert (e3 < e1)
    assert (Rational(170, 10) < e1)
    assert not (Rational(170, 10) > e1)
    assert (e1_prime == e1)
    assert (e2 < e3)
```

File: sympy/core/numbers.py [hunk 7]
```diff
@@ -3413,6 +3416,216 @@ def __hash__(self):
         return super(NumberSymbol, self).__hash__()
 
 
+class BigNumber:
```
Comment (sylee957) [inline sympy/core/numbers.py:3419]: I think that it's just better to inherit `Basic` or `Expr` if you expect this to work with sympy operators.

**Comment by oscarbenjamin** [comment]
I don't think that this class is needed. We already have `Pow(10, 10, evaluate=False)` and `UnevaluatedExpr` so we don't need a special class for this. If there is some reason why `evaluate=False` it `UnevaluatedExpr` does not work for you then maybe that should be fixed but I don't think having an alternative class that reimplements all of the logic helps.

**Comment by seadavis** [comment]
@oscarbenjamin If you go, UnevaluatedExpr(Pow(10, 10)),  it will evaluate to 1000. Furthermore if you go, UnevaluatedExpr(Pow(10, Pow(10, Pow(10, 100, evaluate=False), evaluate=False), evaluate=False)) > 100, it evaluates to 10**(10**(10**100)) > 100. Whereas BigNumber would evaluate to try. That's BigNumbers goal, is allow for comparison of exteremly large numbers. I don't see how UnevaluatedExpr allows for such a comparison to take place. However, I do get your point about operations being handled properly. So, maybe BigNumber ought to inherit from UnevaluatedExpr? @sylee957, do you have an opinion on BigNumber inheriting from UnevaluatedExpr vs, Basic or Expr?

File: sympy/core/numbers.py (hunk 7 above)
Comment (seadavis) [reply sympy/core/numbers.py:3419]: ok, that makes sense to me. I'll inherit from Expr, I think. Or maybe unevaluated expression?

**Comment by oscarbenjamin** [comment]
You should use it like:
```
In [5]: UnevaluatedExpr(10)**10                                                                                                   
Out[5]: 
  10
10  
```
I don't see why this PR is adding a new class. If `UnevaluatedExpr` or `evaluate=False` does not do what you want then let's fix that problem.

**Comment by sylee957** [comment]
I also think that it's better to work on the original `Pow` object with `evaluate=False` because it's not unique to add a new number class for this.

If you are going to add a new class, then I think that it's better to have stuff like `Tetration` that can be designed to work with large numbers symbolically. Tetration can be more specific and can be unique for this purpose.

**Comment by oscarbenjamin** [comment]
I think that very large Integer/Float expressions should not evaluate by default. Again we don't need a new class for that: we should make it happen for the existing Integer/Float classes. There are a number of open issues about this.

**Comment by seadavis** [comment]
Ok, I see what you're saying. I'll  move the comparison logic I have in BigNumber into Pow, and eliminate the BigNumber class.

I can also look into not evaluating large Integer, Float expressions.

**Comment by oscarbenjamin** [comment]
If you dig into this you'll find that actually what we need is #18778

**Comment by seadavis** [comment]
@oscarbenjamin it looks as though, you are quite right, and to even get unevaluated pows to compare symbolically I'll need #18778. It looks a little contreversial? If I  do as the issue says does it have a high chance of getting merged?

**Comment by oscarbenjamin** [comment]
I think it has a high chance of getting merged (if it is done well).

**Comment by czgdp1807** [comment]
@seadavis Are you working on it? This is a good work and can be merged once completed. Thanks for contributions.

**Comment by seadavis** [comment]
@czgdp1807 Yes, but I do need PR #19107 (https://github.com/sympy/sympy/pull/19107) to be finished before I can continue. That PR I will finish tonight,

**Pull request closed without merging**